spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -150,8 +167,30 @@ else if (reply instanceof AbstractIntegrationMessageBuilder) {
 			}
 		}
 
-		Message<?> replyMessage = createOutputMessage(reply, requestHeaders);
-		sendOutput(replyMessage, replyChannel);
+		if (this.asyncReplySupported && reply instanceof ListenableFuture<?>) {
+			ListenableFuture<?> future = (ListenableFuture<?>) reply;
+			final Object fallbackReplyChannel = replyChannel;
+			future.addCallback(new ListenableFutureCallback<Object>() {
+
+				@Override
+				public void onSuccess(Object result) {
+					sendOutput(createOutputMessage(result, requestHeaders), fallbackReplyChannel);
+				}
+
+				@Override
+				public void onFailure(Throwable ex) {
+					Throwable result = ex;
+					if (!(ex instanceof MessagingException)) {
+						result = new MessageHandlingException(requestMessage, ex);
+					}
+					sendOutput(createOutputMessage(result, requestHeaders), fallbackReplyChannel);
```

Comment:
Looks like it would be better to send the error to the `requestHeaders.getErrorChannel()`. Isn't it?